Enable safe HTML rendering for cart item data

ralucaStan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
This pull request introduces several improvements to how product item_data details are rendered and sanitized in the WooCommerce Cart, Checkout and Mini-cart blocks, focusing on safer handling of HTML and better compatibility with the interactivity API. 

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes  https://linear.app/a8c/issue/WOOPLUG-5681/enable-safe-html-rendering-in-block-cartcheckout-metadata

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->
#### Mini-Cart

| Before | After |
| ------ | ----- |
|     <img width="456" height="525" alt="Screenshot 2025-10-15 at 10 57 04" src="https://github.com/user-attachments/assets/37618ce6-6115-4e13-bd5f-3ac76a09439f" /> |  <img width="457" height="588" alt="Screenshot 2025-10-15 at 10 54 41" src="https://github.com/user-attachments/assets/46c70194-48c3-4f4a-9c2b-2dc8d641671c" />     |

#### Cart block 

| Before | After |
| ------ | ----- |
|     <img width="855" height="719" alt="Screenshot 2025-10-15 at 10 57 19" src="https://github.com/user-attachments/assets/f6d8d350-accd-44f1-8628-091a1387b8a1" />   | <img width="828" height="738" alt="Screenshot 2025-10-15 at 10 54 27" src="https://github.com/user-attachments/assets/13ae3705-e516-4c6d-859c-b95688617ba0" />     |

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

Add this code snippet that adds some custom item data:

```
add_filter( 'woocommerce_get_item_data', 'add_extra_item_data_attributes', 10, 2 );

function add_extra_item_data_attributes($item_data, $cart_item) {

    $item_data[] = array(
        'name'  => __('Your <b>Gift Message</b>', 'your-textdomain'),
        'value' => sprintf(
            '<a href="https://www.woocommerce.com" target="_blank" rel="noopener noreferrer" extra="do not display">Click & see</a>'
        ),
    );

    return $item_data;
}
```

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add a simple product and a variable product to cart
2. Go to the Cart, Mini-Cart and Checkout block and see that the HTML tags are rendered , only with the allowed attributes




<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ralucaStan committed:
Use wp_kses_post on Store API cart item data fields

ralucaStan committed:
Sanitize cart item item data in the cart & checkout blocks

ralucaStan committed:
Use sanitized cart item data in the Mini Cart block
Also, display variation data after item data, similar to the other checkout blocks.

ralucaStan committed:
Add changelog

ralucaStan committed:
Fix linting issues

ralucaStan committed:
Update ProductDetails tests and implementation

ralucaStan committed:
Add back the semi-colon

ralucaStan committed:
Add semi-colon to the mini cart block

ralucaStan committed:
Merge branch 'trunk' into wooprd-956-enable-safe-html-rendering-in-block-cartcheckout-metadata

ralucaStan committed:
Remove block tags and improve classname

ralucaStan committed:
Merge branch 'trunk' into wooprd-956-enable-safe-html-rendering-in-block-cartcheckout-metadata

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -701,7 +703,7 @@ const { state: cartItemState } = store(
 				// Decode entities.
 				const nameTxt = document.createElement( 'textarea' );
-				nameTxt.innerHTML = dataItemAttrKey + ':';
+				nameTxt.innerHTML = dataItemAttrKey;
```
Moved this part to the php file

ralucaStan commented on plugins/woocommerce/src/Blocks/BlockTypes/MiniCartProductsTableBlock.php:
```diff
@@ -176,8 +176,10 @@ class="wc-block-formatted-money-amount wc-block-components-formatted-money-amoun
+											<?php echo $this->render_experimental_iapi_product_details_markup( 'item_data' ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>
+										</div>
 										<?php echo $this->render_experimental_iapi_product_details_markup( 'variation' ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>
-										<?php echo $this->render_experimental_iapi_product_details_markup( 'item_data' ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>
```
In the other blocks the itemData is displayed before variations

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/product-details/index.tsx:
```diff
@@ -40,28 +69,50 @@ const ProductDetails = ( {
+				const tempDiv = document.createElement( 'div' );
+				tempDiv.innerHTML = name;
+				const nameForClass =
+					tempDiv.textContent || tempDiv.innerText || '';
```
keeping the classes backwards compatible, and rendering only the text.

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -850,6 +852,77 @@ const { state: cartItemState } = store(
+						'div span.wc-block-components-product-details__value'
+					);
+
+					const name = item?.key || item.name || '';
```
This was aligned this with the Cart & Checkout blocks logic

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/product-details/index.tsx:
```diff
@@ -2,17 +2,46 @@
+	'li',
+	'img',
+	'span',
+	'ol',
```
only allowing inline tags

senadir replied:
Still going through the review but styling can still make them blocks, are you trying to prevent that?

ralucaStan replied:
I cared about valid HTML syntax, and so we prevented having block elements inside inline elements. Styling doesn't impact the HTML validity.

LuigiPulcini replied:
`ul` and `ol` are block-level elements and they could be useful inside the meta value. Wouldn't it be more flexible to change the outer `<span>` tags for key and value to `<div>` and allow block-level tags inside them?

ralucaStan replied:
@LuigiPulcini, you are correct, those are block-level elements, and should be removed. I missed doing that.

> Wouldn't it be more flexible to change the outer <span> tags for key and value to <div> and allow block-level tags inside them?

This is one option, but I am not sure what impact it will have on existing stores. 

I would rather bring in this change outside of this PR, and remove the `ol` and `li` elements from the allowed tags. Not saying it won't happen in the future, I'm just keeping the scope of the PR limited.

ralucaStan replied:
Also, the change isn't aligned with the intention behind this area, which is to show metadata. So, if some plugins are using this for more complex use cases, we might need alternative solutions.

LuigiPulcini replied:
Thanks, @ralucaStan. I totally understand the need to keep the scope as limited as possible, particularly with backward compatibility in mind. I only want to point out that non-scalar meta values are not such an extravagant use case. In that situation, limiting the allowed tags to in-line elements has the potential of creating semantic issues – especially when it comes to accessibility.

ralucaStan replied:
> I only want to point out that non-scalar meta values are not such an extravagant use case. In that situation, limiting the allowed tags to in-line elements has the potential of creating semantic issues – especially when it comes to accessibility.

Could you provide some examples? What semantic issues do you have in mind?

LuigiPulcini replied:
Lists in general would be the most obvious example. If the meta value contains an array of image URLs, listing them in the metadata would be tricky, since both the following markups would be semantically incorrect to represent a list of images (or anchors for that matter):

```html
<span>
    <img src="image1">
    <img src="image2">
    <img src="image3">
</span>
```

```html
<span>
    <span><img src="image1"></span>
    <span><img src="image2"></span>
    <span><img src="image3"></span>
</span>
```

Again, I totally understand the need to keep the scope of this PR as limited as possible. But the risk is to make a change that is more cosmetic than functional.

ralucaStan replied:
@LuigiPulcini, we've looked at your use case and decided not to allow images at this level. The Product team is aware of the use case and will consider it in the extensibility strategy. 

Some reasons:
- The item_data is displayed in Mini-Cart, Cart and Checkout and it's impacting the layout, the Mini-Cart being the main concern
- Adding the images and loading separate scripts for a slider is not an ideal usage for this area. If this is a need, the blocks would have to control the display of those images. Think there are different plugins with different sliders

LuigiPulcini replied:
@ralucaStan Thanks for letting me know. I must admit this is rather disappointing. I don't understand the rationale behind this choice, but I can only accept your decision.

I guess we will keep recommending our customers to replace the cart and checkout blocks with the classic cart and classic checkout ones.

LuigiPulcini replied:
Essentially, by looking at the array of allowed tags, these changes were all aimed at supporting only anchors and a bunch of formatting tags. I am not sure it was really worth the effort.

ralucaStan committed:
remove style attribute

ralucaStan committed:
improve types for cart item data in the iapi store

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/product-details/index.tsx:
```diff
@@ -2,17 +2,46 @@
+
+const CONTENT_ATTR = [
+	'target',
+	'href',
+	'rel',
+	'name',
+	'download',
+	'src',
+	'style',
+	'class',
+	'title',
+];
```
removed, let's start without it

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/product-details/index.tsx:
```diff
@@ -40,28 +68,50 @@ const ProductDetails = ( {
 			{ details.map( ( detail ) => {
 				// Support both `key` and `name` props
 				const name = detail?.key || detail.name || '';
+				// Strip HTML tags from name for CSS class generation
+				const tempDiv = document.createElement( 'div' );
+				tempDiv.innerHTML = name;
+				const nameForClass =
+					tempDiv.textContent || tempDiv.innerText || '';
```
```
	const sanitizedName = sanitizeHTML( name, {
tags: [],
attr: [],
} );
```
I already considered this one, but it still keeps HTML tags like `b `

senadir replied:
Do we know if the attribute name or key is already sanitized server side? I think it is, so this shouldn't be a problem.

ralucaStan replied:
Variation data is [sanitized here](https://github.com/woocommerce/woocommerce/blob/b9247efc730dd1bb1e20616ec813acf9cab05bd5/plugins/woocommerce/src/StoreApi/Schemas/V1/CartItemSchema.php#L70-L71), but we're also sanitizing  `item_data` which comes from filters. I'll have a look at the suggested approach.

ralucaStan replied:
`item_data` comes sanitized, so no harmful tags are present there. But we need to strip all HTML content and only keep the text to have backwards compatible CSS classes

senadir reviewed (approved):
This is testing fine for me when it comes to code, but I'm not sure how much freedom/guardrails we want to offer.

For now we accept both classnames and inline CSS, which means you can style this however you like, do we care about allowing this?

<img width="500" height="520" alt="Image" src="https://github.com/user-attachments/assets/e2cb6465-4e3a-4290-b391-45770d848cdf" />

ralucaStan committed:
Remove list related block elements from allowed tags

samueljseay commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -850,6 +846,77 @@ const { state: cartItemState } = store(
+					item: CartItemDataAttr
+				) => {
+					const nameEl = element.querySelector(
+						'div span.wc-block-components-product-details__name'
```
Is it possible to have watch run on the element you intend to inject the content?

samueljseay commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -850,6 +846,77 @@ const { state: cartItemState } = store(
+					return;
+				}
+
+				// A workaround for the lack of dangerous set HTML directive
```
As mentioned before, while we do need this workaround, I think it'd be much better to try isolate changes to the same element being watched (ideally some internal span) to avoid what becomes a lot of imperative DOM updates, rather than what should be (imo) an absolutely minimal setting innerHTML on the ref itself.

samueljseay commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -850,6 +846,77 @@ const { state: cartItemState } = store(
 					}
 				}
 			},
+			itemItemData() {
```
Looking over this I think the iAPI code can be simplified.

My feeling is that I'd add 2 directives like singleItemNameInnerHTML and singleItemValueInnerHTML and minimize both imperative DOM querying/manipulation and duplication of the deriving of the item name and value. I'm assuming that you'd add a watch directive on each li element directly.

I'd add watch directly on the node that should have the innerHTML set, and then I'd calculate the name or value. the inner functions should be removed.

write 2 derived state getters one for name and one for value that duplicate the logic from React.

pseudo code:

```
get cartItemNameHTML() {
  const itemData = cartItemState.cartItem?.item_data || [];
  const name = item?.key || item.name || '';

  return trimWords(name);
}

get cartItemValueHTML() {
  ... same as name
}

...
// watch directive
singleItemNameInnerHTML() {
  const { ref } = getElement();

  if ( ! ref ) {
   return;
  }

  ref.innerHTML = state.cartItemNameHTML;
}

singleItemValueInnerHTML() {
  // much the same
}
```

Even if for some reason this isn't possible, you should be able to take some concepts from this example such as derived state getter to deduplicate logic and simplify the implementation.

I also hope if its possible to populate initial values with wp_kses_post server side that HTML updating would also be minimized.

ralucaStan replied:
Thanks Sam, great feedback. This was the 1st time I wrote iAPI and so the`ref` wasn't obvious to me. 
I've added the new directives to the `name` and `value` `span` elements, and that made the implementation much cleaner.

I didn't add new getters because the item value is actually coming from the `cartItemDataAttr` getter.

I've added small improvements to this getter to make it obvious that it's handling `item_data` and `variation` metadata. IMO the current implementation makes the getter harder to understand, and we could discuss improving it in the future by having 2 separate getters. @SantosGuillamot for awareness

samueljseay commented on plugins/woocommerce/src/Blocks/BlockTypes/MiniCartProductsTableBlock.php:
```diff
@@ -289,18 +291,19 @@ class="wc-block-components-product-details"
 		>
-			<span class="wc-block-components-product-details__name" data-wp-text="state.cartItemDataAttr.name"></span>
-			<span class="wc-block-components-product-details__value" data-wp-text="state.cartItemDataAttr.value"></span>
+			<span class="wc-block-components-product-details__name"<?php echo $skip_text ? '' : ' data-wp-text="state.cartItemDataAttr.name"'; ?>></span><?php echo ': '; ?>
```
If we skip the text here it won't be rendered initially via server rendering which works right now as it is.

This is quite a downside of this approach. I would have thought we could populate the server state with a value that has had wp_kses_post run over it for the initial render as opposed to rendering it blank.

Also not sure if its a concern, but the `:` is always printed even if there is no value or name, I'm not sure if either is possible but the React logic implies it could be.

ralucaStan replied:
This function is responsible for rendering both `variation` information, from a Variable product, and `item_data` which is coming from plugings.

I've changed the output depending on the type, because the `item_data` should be sanitized.

ralucaStan replied:
When we render the item there's always a `name` and a `value`.

ralucaStan replied:
> Also not sure if its a concern, but the : is always printed even if there is no value or name, I'm not sure if either is possible but the React logic implies it could be.


I followed the blocks approach here, but you are right, it doesn't hurt to check

samueljseay commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -850,6 +846,77 @@ const { state: cartItemState } = store(
+						'div span.wc-block-components-product-details__value'
+					);
+
+					const name = item?.key || item.name || '';
```
I think the easiest win here as I mentioned before would be derived getters for these so this logic is not duplicated.

ralucaStan replied:
I've moved this logic to `cartItemDataAttr`, where the metadata item is constructed.

samueljseay reviewed (changes requested):
I think some of the implementation can be simplified. I have left some suggestions on what could be improved.

ralucaStan committed:
Improve the handling of sanitizing the item data
- improved the cartItemDataAttr to understand that it handles both item_data and variations, aligned class to the other blocks
- added separate callbacks for name and value
- php function renders item markup depending on the item type: variations are safe so we output them, and item_data is added via callback

samueljseay commented on plugins/woocommerce/src/Blocks/BlockTypes/MiniCartProductsTableBlock.php:
```diff
@@ -289,18 +292,25 @@ class="wc-block-components-product-details"
+
 		>
-			<span class="wc-block-components-product-details__name" data-wp-text="state.cartItemDataAttr.name"></span>
+		<?php if ( $is_item_data ) : ?>
+			<span class="wc-block-components-product-details__name" data-wp-watch="callbacks.itemDataNameInnerHTML"></span><?php echo ': '; ?>
+			<span class="wc-block-components-product-details__value" data-wp-watch="callbacks.itemDataValueInnerHTML"></span>
+		<?php else : ?>
+			<span class="wc-block-components-product-details__name"  data-wp-text="state.cartItemDataAttr.name"></span><?php echo ': '; ?>
 			<span class="wc-block-components-product-details__value" data-wp-text="state.cartItemDataAttr.value"></span>
+		<?php endif; ?>
```
@ralucaStan It looks like this is still revelant. I was hoping if possible that the item data could be populated in server side state to avoid it being empty on first render

ralucaStan replied:
We are in a template here where we're using the cart state, not the cart session. The Mini-Cart is closed on the 1st render, so I don't think this is an issue. The `short_description` also comes as an empty div.

Presuming some `item_data` comes without tags, we could still use the text directive to render the values, which for some products might come with some HTML tags, but then it will be replace via the callback. 

Let me know if I'm missing something.

samueljseay replied:
>  The Mini-Cart is closed on the 1st render, so I don't think this is an issue

Ah yes, thats a good point! Yep that minimizes any problem I think :)

luisherranz commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -850,6 +852,32 @@ const { state: cartItemState } = store(
+				if ( ! ref ) {
+					return;
+				}
+
+				const dataAttr = cartItemState.cartItemDataAttr;
+				if ( 'value' in dataAttr && dataAttr.value ) {
+					ref.innerHTML = trimWords( dataAttr.value );
+				}
+			},
```
We can simplify these two callbacks so that they both call the same function:

```js
function itemDataInnerHTML( field ) {
  const { ref } = getElement();
  // ...
}

// ...
callbacks: {
  itemDataNameInnerHTML() {
    itemDataInnerHTML( 'name' );
  },
  itemDataValueInnerHTML() {
    itemDataInnerHTML( 'value' );
  },
  // ...
}
```

Or use an attribute to pass an argument to the callback.

```html
<span data-wc-field="name" data-wp-watch="callbacks.itemDataInnerHTML"></span>
<span data-wc-field="value" data-wp-watch="callbacks.itemDataInnerHTML"></span>
```

```js
callbacks: {
  itemDataInnerHTML() {
    const { ref, attributes } = getElement();
    const field = attributes[ 'data-wc-field' ];
    // ...
  },
}
```

ralucaStan replied:
Thanks Luis, I went with the first proposal, as we don't really use `data-wc-..` attributes in other places.

ralucaStan committed:
Avoid repetition by adding itemDataInnerHTML

ralucaStan committed:
Prevent image tags in the metadata fields
The item_data gets render across the Mini-Cart, Checkout and Cart. Allowing images at this level impact the overall UX of the checkout

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/blocks/mini-cart/iapi-frontend.ts:
```diff
@@ -30,6 +30,29 @@ import { translateJQueryEventToNative } from '../../base/stores/woocommerce/lega
+			ref.innerHTML = trimWords( value );
+		}
+	}
+}
```
The content is already sanitized, and sanitizeHTML can't be used in the iAPI context

ralucaStan committed:
Merge branch 'trunk' into wooprd-956-enable-safe-html-rendering-in-block-cartcheckout-metadata

samueljseay reviewed (approved):
From iAPI perspective I think this is looking good now. I have tested and it tests well for me :)

ralucaStan committed:
Merge branch 'trunk' into wooprd-956-enable-safe-html-rendering-in-block-cartcheckout-metadata

ralucaStan committed:
Move itemDataInnerHTML function

ralucaStan committed:
Merge branch 'wooprd-956-enable-safe-html-rendering-in-block-cartcheckout-metadata' of github.com:woocommerce/woocommerce into wooprd-956-enable-safe-html-rendering-in-block-cartcheckout-metadata

ralucaStan committed:
Merge branch 'trunk' into wooprd-956-enable-safe-html-rendering-in-block-cartcheckout-metadata

ralucaStan committed:
Merge branch 'trunk' into wooprd-956-enable-safe-html-rendering-in-block-cartcheckout-metadata

ralucaStan committed:
Merge branch 'trunk' into wooprd-956-enable-safe-html-rendering-in-block-cartcheckout-metadata
